torch/csrc/autograd/python_function.cpp

```diff
@@ -212,30 +212,47 @@ static int THPFunction_traverse(THPFunction* self, visitproc visit, void* arg) {
   //
   // TODO: I'm not really sure if we're actually obligated to traverse PyObject
   // that is stored in PyNode, since we don't really own that C++ object.
-  if (auto cdata = self->cdata.lock()) {
-    for (const auto& hook : cdata->tensor_pre_hooks()) {
-      if (auto pyhook = dynamic_cast<PyFunctionTensorPreHook*>(hook.get())) {
-        Py_VISIT(pyhook->dict);
+  // Take care to keep the call to .lock() in a scope, so that we release our
+  // shared pointer before traversing any saved variables.
+  // This edge case can come up if any saved variables (indirectly) hold
+  // a strong reference to cdata (our PyNode).
+  {
+    if (auto cdata = self->cdata.lock()) {
```

Comment:
This required some care. our python custom function `THPFunction` holds a weak pointer to its autograd node, `self->cdata` (it is of type `PyNode`). This very same `PyNode` is part of the reference cycle (see the diagram I put in the description). The piece of code here temporarily takes out a strong reference to that PyNode in order to visit any hooks on it, etc. But when we traverse the saved variables later, I need to make sure that the strong reference is gone.